Failed to create VOLUME from snapshot

<!--
Verify first that your issue/request is not already reported on GitHub.
Also test if the latest release and master branch are affected too.
Always add information AFTER of these HTML comments, but no need to delete the comments.
-->

##### ISSUE TYPE
<!-- Pick one below and delete the rest -->
 * Bug Report


##### COMPONENT NAME
<!--
Categorize the issue, e.g. API, VR, VPN, UI, etc.
-->
~~~
snapshot
~~~

##### CLOUDSTACK VERSION
<!--
New line separated list of affected versions, commit ID for issues on master branch.
-->

~~~
4.14.0.0
~~~

##### CONFIGURATION
<!--
Information about the configuration if relevant, e.g. basic network, advanced networking, etc.  N/A otherwise
-->


##### OS / ENVIRONMENT
<!--
Information about the environment if relevant, N/A otherwise
-->
CentOS 7.7 KVM QEMU 2.12.0,libvirt 4.5.0
Ceph 14.2.10

##### SUMMARY
<!-- Explain the problem/feature briefly -->


snapshot.backup.to.secondary=false
kvm.snapshot.enabled=true


When creating a new VOLUME through a VOLUME snapshot,
Only the first snapshot of each VOLUME is successful;
No other snapshots of the same VOLUME can create a new VOLUME

"Path" is missing from CopyCommand: "snapshots/2/5445/4e130b18-a495-4220-8c8d-118d186e961f"

CopyCommand:
{
	"org.apache.cloudstack.storage.command.CopyCommand": {
		"srcTO": {
			"org.apache.cloudstack.storage.to.SnapshotObjectTO": {
				"volume": {
					"uuid": "8c39f2f5-9b18-494e-b990-969d9318eeff",
					"volumeType": "ROOT",
					"dataStore": {
						"org.apache.cloudstack.storage.to.PrimaryDataStoreTO": {
							"uuid": "049f30ad-d4ec-33cb-a0e3-e34df5f2f757",
							"id": 2,
							"poolType": "RBD",
							"host": "10.226.15.3,10.226.15.4,10.226.15.5",
							"path": "cloudstack",
							"port": 6789,
							"url": "RBD://10.226.15.3,10.226.15.4,10.226.15.5/cloudstack/?ROLE=Primary&STOREUUID=049f30ad-d4ec-33cb-a0e3-e34df5f2f757",
							"isManaged": false
						}
					},
					"name": "ROOT-4158",
					"size": 107374182400,
					"path": "8c39f2f5-9b18-494e-b990-969d9318eeff",
					"volumeId": 5445,
					"vmName": "i-2-4158-VM",
					"accountId": 2,
					"format": "QCOW2",
					"provisioningType": "THIN",
					"id": 5445,
					"deviceId": 0,
					"hypervisorType": "KVM",
					"directDownload": false
				},
				"dataStore": {
					"com.cloud.agent.api.to.NfsTO": {
						"_url": "nfs://10.226.15.21/nfs/secstorage",
						"_role": "Image"
					}
				},
				"vmName": "i-2-4158-VM",
				"name": "aaaaa2",
				"hypervisorType": "KVM",
				"id": 27345,
				"quiescevm": false,
				"physicalSize": 0
			}
		},
		"destTO": {
			"org.apache.cloudstack.storage.to.VolumeObjectTO": {
				"uuid": "32cc7365-9bc3-49c5-87f1-871a96fb0bf7",
				"volumeType": "DATADISK",
				"dataStore": {
					"org.apache.cloudstack.storage.to.PrimaryDataStoreTO": {
						"uuid": "049f30ad-d4ec-33cb-a0e3-e34df5f2f757",
						"id": 2,
						"poolType": "RBD",
						"host": "10.226.15.3,10.226.15.4,10.226.15.5",
						"path": "cloudstack",
						"port": 6789,
						"url": "RBD://10.226.15.3,10.226.15.4,10.226.15.5/cloudstack/?ROLE=Primary&STOREUUID=049f30ad-d4ec-33cb-a0e3-e34df5f2f757",
						"isManaged": false
					}
				},
				"name": "aaaaa2-Volume",
				"size": 107374182400,
				"volumeId": 5447,
				"accountId": 2,
				"format": "QCOW2",
				"provisioningType": "THIN",
				"id": 5447,
				"hypervisorType": "KVM",
				"directDownload": false
			}
		},
		"executeInSequence": false,
		"options": {},
		"options2": {},
		"wait": 10800
	}
}

Log:

2020-10-28 09:03:55,535 WARN  [o.a.c.s.d.ObjectInDataStoreManagerImpl] (API-Job-Executor-1:ctx-d3f6d1cd job-127275 ctx-f465d613) (logid:91387467) Unsupported data object (VOLUME, org.apache.cloudstack.storage.datastore.PrimaryDataStoreImpl@2b441866), no need to delete from object in store ref table
2020-10-28 09:03:55,539 DEBUG [o.a.c.e.o.VolumeOrchestrator] (API-Job-Executor-1:ctx-d3f6d1cd job-127275 ctx-f465d613) (logid:91387467) Failed to create volume from snapshot:java.lang.NullPointerException
	at com.cloud.hypervisor.kvm.storage.KVMStorageProcessor.createVolumeFromSnapshot(KVMStorageProcessor.java:1599)
	at com.cloud.storage.resource.StorageSubsystemCommandHandlerBase.execute(StorageSubsystemCommandHandlerBase.java:110)
	at com.cloud.storage.resource.StorageSubsystemCommandHandlerBase.handleStorageCommands(StorageSubsystemCommandHandlerBase.java:57)
	at com.cloud.hypervisor.kvm.resource.wrapper.LibvirtStorageSubSystemCommandWrapper.execute(LibvirtStorageSubSystemCommandWrapper.java:36)
	at com.cloud.hypervisor.kvm.resource.wrapper.LibvirtStorageSubSystemCommandWrapper.execute(LibvirtStorageSubSystemCommandWrapper.java:30)
	at com.cloud.hypervisor.kvm.resource.wrapper.LibvirtRequestWrapper.execute(LibvirtRequestWrapper.java:89)
	at com.cloud.hypervisor.kvm.resource.LibvirtComputingResource.executeRequest(LibvirtComputingResource.java:1665)
	at com.cloud.agent.Agent.processRequest(Agent.java:644)
	at com.cloud.agent.Agent$AgentRequestHandler.doTask(Agent.java:1069)
	at com.cloud.utils.nio.Task.call(Task.java:83)
	at com.cloud.utils.nio.Task.call(Task.java:29)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)

##### STEPS TO REPRODUCE
<!--
For bugs, show exactly how to reproduce the problem, using a minimal test-case. Use Screenshots if accurate.

For new features, show how the feature would be used.
-->

<!-- Paste example playbooks or commands between quotes below -->
~~~
1. Create a VOLUME snapshot;
2. Create VOLUME from the volume snapshot (successful);
3. Create a VOLUME snapshot;
4. Create VOLUME from volume snapshot (failure);

~~~

<!-- You can also paste gist.github.com links for larger files -->

##### EXPECTED RESULTS
<!-- What did you expect to happen when running the steps above? -->

~~~
1. Create a VOLUME snapshot;
2. Create VOLUME from the volume snapshot (successful);
3. Create a VOLUME snapshot;
4. Create VOLUME from volume snapshot (successful);
~~~

##### ACTUAL RESULTS
<!-- What actually happened? -->

<!-- Paste verbatim command output between quotes below -->
~~~

~~~

I was able to reproduce on NFS storage pool, and the same setting values.

```
2020-10-28 15:35:59,214 DEBUG [c.c.h.o.r.Ovm3HypervisorGuru] (API-Job-Executor-32:ctx-ee634337 job-44 ctx-540c9792) (logid:8bbe9b7f) getCommandHostDelegation: class org.apache.cloudstack.storage.command.CopyCommand
2020-10-28 15:35:59,215 DEBUG [c.c.a.t.Request] (API-Job-Executor-32:ctx-ee634337 job-44 ctx-540c9792) (logid:8bbe9b7f) Seq 1-6719370644036780103: Sending  { Cmd , MgmtId: 32988956015047, via: 1(ref-trl-1821-k-M7-nicolas-vazquez-kvm1), Ver: v1, Flags: 100011, [{"org.apache.cloudstack.storage.command.CopyCommand":{"srcTO":{"org.apache.cloudstack.storage.to.SnapshotObjectTO":{"volume":{"uuid":"88a6ce2c-de4c-475f-827e-696f0f0bdeb2","volumeType":"ROOT","dataStore":{"org.apache.cloudstack.storage.to.PrimaryDataStoreTO":{"uuid":"9715ad47-244e-3d24-a382-7ad33a3dc0f1","id":1,"poolType":"NetworkFilesystem","host":"10.10.0.16","path":"/acs/primary/ref-trl-1821-k-M7-nicolas-vazquez/ref-trl-1821-k-M7-nicolas-vazquez-kvm-pri1","port":2049,"url":"NetworkFilesystem://10.10.0.16/acs/primary/ref-trl-1821-k-M7-nicolas-vazquez/ref-trl-1821-k-M7-nicolas-vazquez-kvm-pri1/?ROLE=Primary&STOREUUID=9715ad47-244e-3d24-a382-7ad33a3dc0f1","isManaged":false}},"name":"ROOT-4","size":8589934592,"path":"88a6ce2c-de4c-475f-827e-696f0f0bdeb2","volumeId":4,"vmName":"i-2-4-VM","accountId":2,"format":"QCOW2","provisioningType":"THIN","id":4,"deviceId":0,"hypervisorType":"KVM","directDownload":false},"dataStore":{"com.cloud.agent.api.to.NfsTO":{"_url":"NFS://10.10.0.16/acs/secondary/ref-trl-1821-k-M7-nicolas-vazquez/ref-trl-1821-k-M7-nicolas-vazquez-sec1","_role":"Image"}},"vmName":"i-2-4-VM","name":"snap2","hypervisorType":"KVM","id":2,"quiescevm":false,"physicalSize":0}},"destTO":{"org.apache.cloudstack.storage.to.VolumeObjectTO":{"uuid":"81832bc9-1c46-4db8-968d-198064a4cec3","volumeType":"DATADISK","dataStore":{"org.apache.cloudstack.storage.to.PrimaryDataStoreTO":{"uuid":"9715ad47-244e-3d24-a382-7ad33a3dc0f1","id":1,"poolType":"NetworkFilesystem","host":"10.10.0.16","path":"/acs/primary/ref-trl-1821-k-M7-nicolas-vazquez/ref-trl-1821-k-M7-nicolas-vazquez-kvm-pri1","port":2049,"url":"NetworkFilesystem://10.10.0.16/acs/primary/ref-trl-1821-k-M7-nicolas-vazquez/ref-trl-1821-k-M7-nicolas-vazquez-kvm-pri1/?ROLE=Primary&STOREUUID=9715ad47-244e-3d24-a382-7ad33a3dc0f1","isManaged":false}},"name":"diskfromsnap2","size":21474836480,"volumeId":7,"accountId":2,"format":"QCOW2","provisioningType":"THIN","id":7,"hypervisorType":"KVM","directDownload":false}},"executeInSequence":false,"options":{},"options2":{},"wait":10800}}] }
2020-10-28 15:35:59,224 DEBUG [c.c.a.t.Request] (AgentManager-Handler-15:null) (logid:) Seq 1-6719370644036780103: Processing:  { Ans: , MgmtId: 32988956015047, via: 1, Ver: v1, Flags: 10, [{"com.cloud.agent.api.Answer":{"result":false,"details":"java.lang.NullPointerException\n\tat com.cloud.hypervisor.kvm.storage.KVMStorageProcessor.createVolumeFromSnapshot(KVMStorageProcessor.java:1599)\n\tat com.cloud.storage.resource.StorageSubsystemCommandHandlerBase.execute(StorageSubsystemCommandHandlerBase.java:110)\n\tat com.cloud.storage.resource.StorageSubsystemCommandHandlerBase.handleStorageCommands(StorageSubsystemCommandHandlerBase.java:57)\n\tat com.cloud.hypervisor.kvm.resource.wrapper.LibvirtStorageSubSystemCommandWrapper.execute(LibvirtStorageSubSystemCommandWrapper.java:36)\n\tat com.cloud.hypervisor.kvm.resource.wrapper.LibvirtStorageSubSystemCommandWrapper.execute(LibvirtStorageSubSystemCommandWrapper.java:30)\n\tat com.cloud.hypervisor.kvm.resource.wrapper.LibvirtRequestWrapper.execute(LibvirtRequestWrapper.java:78)\n\tat com.cloud.hypervisor.kvm.resource.LibvirtComputingResource.executeRequest(LibvirtComputingResource.java:1637)\n\tat com.cloud.agent.Agent.processRequest(Agent.java:644)\n\tat com.cloud.agent.Agent$AgentRequestHandler.doTask(Agent.java:1057)\n\tat com.cloud.utils.nio.Task.call(Task.java:83)\n\tat com.cloud.utils.nio.Task.call(Task.java:29)\n\tat java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\tat java.base/java.lang.Thread.run(Thread.java:834)\n","wait":0}}] }
2020-10-28 15:35:59,225 DEBUG [c.c.a.t.Request] (API-Job-Executor-32:ctx-ee634337 job-44 ctx-540c9792) (logid:8bbe9b7f) Seq 1-6719370644036780103: Received:  { Ans: , MgmtId: 32988956015047, via: 1(ref-trl-1821-k-M7-nicolas-vazquez-kvm1), Ver: v1, Flags: 10, { Answer } }
2020-10-28 15:35:59,232 WARN  [o.a.c.s.d.ObjectInDataStoreManagerImpl] (API-Job-Executor-32:ctx-ee634337 job-44 ctx-540c9792) (logid:8bbe9b7f) Unsupported data object (VOLUME, org.apache.cloudstack.storage.datastore.PrimaryDataStoreImpl@6689d2a6), no need to delete from object in store ref table
2020-10-28 15:35:59,237 DEBUG [o.a.c.e.o.VolumeOrchestrator] (API-Job-Executor-32:ctx-ee634337 job-44 ctx-540c9792) (logid:8bbe9b7f) Failed to create volume from snapshot:java.lang.NullPointerException
	at com.cloud.hypervisor.kvm.storage.KVMStorageProcessor.createVolumeFromSnapshot(KVMStorageProcessor.java:1599)
	at com.cloud.storage.resource.StorageSubsystemCommandHandlerBase.execute(StorageSubsystemCommandHandlerBase.java:110)
	at com.cloud.storage.resource.StorageSubsystemCommandHandlerBase.handleStorageCommands(StorageSubsystemCommandHandlerBase.java:57)
	at com.cloud.hypervisor.kvm.resource.wrapper.LibvirtStorageSubSystemCommandWrapper.execute(LibvirtStorageSubSystemCommandWrapper.java:36)
	at com.cloud.hypervisor.kvm.resource.wrapper.LibvirtStorageSubSystemCommandWrapper.execute(LibvirtStorageSubSystemCommandWrapper.java:30)
	at com.cloud.hypervisor.kvm.resource.wrapper.LibvirtRequestWrapper.execute(LibvirtRequestWrapper.java:78)
	at com.cloud.hypervisor.kvm.resource.LibvirtComputingResource.executeRequest(LibvirtComputingResource.java:1637)
	at com.cloud.agent.Agent.processRequest(Agent.java:644)
	at com.cloud.agent.Agent$AgentRequestHandler.doTask(Agent.java:1057)
	at com.cloud.utils.nio.Task.call(Task.java:83)
	at com.cloud.utils.nio.Task.call(Task.java:29)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)

2020-10-28 15:35:59,248 DEBUG [c.c.r.ResourceLimitManagerImpl] (API-Job-Executor-32:ctx-ee634337 job-44 ctx-540c9792) (logid:8bbe9b7f) Updating resource Type = volume count for Account = 2 Operation = decreasing Amount = 1
2020-10-28 15:35:59,255 DEBUG [c.c.r.ResourceLimitManagerImpl] (API-Job-Executor-32:ctx-ee634337 job-44 ctx-540c9792) (logid:8bbe9b7f) Updating resource Type = primary_storage count for Account = 2 Operation = decreasing Amount = 21474836480
2020-10-28 15:35:59,269 ERROR [c.c.a.ApiAsyncJobDispatcher] (API-Job-Executor-32:ctx-ee634337 job-44) (logid:8bbe9b7f) Unexpected exception while executing org.apache.cloudstack.api.command.admin.volume.CreateVolumeCmdByAdmin
com.cloud.utils.exception.CloudRuntimeException: Failed to create volume: 7
	at com.cloud.storage.VolumeApiServiceImpl.createVolume(VolumeApiServiceImpl.java:837)
	at com.cloud.storage.VolumeApiServiceImpl.createVolume(VolumeApiServiceImpl.java:184)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.apache.cloudstack.network.contrail.management.EventUtils$EventInterceptor.invoke(EventUtils.java:107)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:175)
	at com.cloud.event.ActionEventInterceptor.invoke(ActionEventInterceptor.java:51)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:175)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:95)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
	at com.sun.proxy.$Proxy212.createVolume(Unknown Source)
	at org.apache.cloudstack.api.command.user.volume.CreateVolumeCmd.execute(CreateVolumeCmd.java:225)
	at com.cloud.api.ApiDispatcher.dispatch(ApiDispatcher.java:156)
	at com.cloud.api.ApiAsyncJobDispatcher.runJob(ApiAsyncJobDispatcher.java:108)
	at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.runInContext(AsyncJobManagerImpl.java:603)
	at org.apache.cloudstack.managed.context.ManagedContextRunnable$1.run(ManagedContextRunnable.java:49)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:56)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:103)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.runWithContext(DefaultManagedContext.java:53)
	at org.apache.cloudstack.managed.context.ManagedContextRunnable.run(ManagedContextRunnable.java:46)
	at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.run(AsyncJobManagerImpl.java:551)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.cloud.utils.exception.CloudRuntimeException: Failed to create volume from snapshot:java.lang.NullPointerException
	at com.cloud.hypervisor.kvm.storage.KVMStorageProcessor.createVolumeFromSnapshot(KVMStorageProcessor.java:1599)
	at com.cloud.storage.resource.StorageSubsystemCommandHandlerBase.execute(StorageSubsystemCommandHandlerBase.java:110)
	at com.cloud.storage.resource.StorageSubsystemCommandHandlerBase.handleStorageCommands(StorageSubsystemCommandHandlerBase.java:57)
	at com.cloud.hypervisor.kvm.resource.wrapper.LibvirtStorageSubSystemCommandWrapper.execute(LibvirtStorageSubSystemCommandWrapper.java:36)
	at com.cloud.hypervisor.kvm.resource.wrapper.LibvirtStorageSubSystemCommandWrapper.execute(LibvirtStorageSubSystemCommandWrapper.java:30)
	at com.cloud.hypervisor.kvm.resource.wrapper.LibvirtRequestWrapper.execute(LibvirtRequestWrapper.java:78)
	at com.cloud.hypervisor.kvm.resource.LibvirtComputingResource.executeRequest(LibvirtComputingResource.java:1637)
	at com.cloud.agent.Agent.processRequest(Agent.java:644)
	at com.cloud.agent.Agent$AgentRequestHandler.doTask(Agent.java:1057)
	at com.cloud.utils.nio.Task.call(Task.java:83)
	at com.cloud.utils.nio.Task.call(Task.java:29)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)

	at org.apache.cloudstack.engine.orchestration.VolumeOrchestrator.createVolumeFromSnapshot(VolumeOrchestrator.java:432)
	at com.cloud.storage.VolumeApiServiceImpl.createVolumeFromSnapshot(VolumeApiServiceImpl.java:859)
	at com.cloud.storage.VolumeApiServiceImpl.createVolume(VolumeApiServiceImpl.java:809)
	... 31 more
2020-10-28 15:35:59,272 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl] (API-Job-Executor-32:ctx-ee634337 job-44) (logid:8bbe9b7f) Complete async job-44, jobStatus: FAILED, resultCode: 530, result: org.apache.cloudstack.api.response.ExceptionResponse/null/{"uuidList":[],"errorcode":530,"errortext":"Failed to create volume: 7"}
```

The `KVMStorageProcessor` is expecting a path but its null.

DB:
![image](https://user-images.githubusercontent.com/5295080/97460956-0050b500-191c-11eb-9bb7-c9449db9cde2.png)

I've created 2 snapshots, both of them have null path but the first worked and second failed

--- a/engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java
+++ b/engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java
@@ -114,6 +114,7 @@ import com.cloud.storage.ScopeType;
 import com.cloud.storage.Snapshot;
 import com.cloud.storage.Storage;
 import com.cloud.storage.Storage.ImageFormat;
+import com.cloud.storage.Storage.StoragePoolType;
 import com.cloud.storage.StorageManager;
 import com.cloud.storage.StoragePool;
 import com.cloud.storage.VMTemplateStorageResourceAssoc;
@@ -124,6 +125,7 @@ import com.cloud.storage.VolumeVO;
 import com.cloud.storage.dao.SnapshotDao;
 import com.cloud.storage.dao.VolumeDao;
 import com.cloud.storage.dao.VolumeDetailsDao;
+import com.cloud.storage.snapshot.SnapshotManager;
 import com.cloud.template.TemplateManager;
 import com.cloud.template.VirtualMachineTemplate;
 import com.cloud.user.Account;
@@ -454,18 +456,25 @@ public class VolumeOrchestrator extends ManagerBase implements VolumeOrchestrati
             if (snapInfo == null) {
                 throw new CloudRuntimeException("Cannot find snapshot " + snapshot.getId());
             }
-            // We need to copy the snapshot onto secondary.
-            SnapshotStrategy snapshotStrategy = _storageStrategyFactory.getSnapshotStrategy(snapshot, SnapshotOperation.BACKUP);
-            snapshotStrategy.backupSnapshot(snapInfo);
 
-            // Attempt to grab it again.
-            snapInfo = snapshotFactory.getSnapshot(snapshot.getId(), dataStoreRole);
-            if (snapInfo == null) {
-                throw new CloudRuntimeException("Cannot find snapshot " + snapshot.getId() + " on secondary and could not create backup");
+            boolean backupSnapToSecondary = SnapshotManager.BackupSnapshotAfterTakingSnapshot.value() == null || SnapshotManager.BackupSnapshotAfterTakingSnapshot.value();
+
+            StoragePoolVO srcPool = _storagePoolDao.findById(snapInfo.getBaseVolume().getPoolId());
+            // We need to copy the snapshot onto secondary.
+            //Skipping the backup to secondary storage with NFS/FS could be supported when CLOUDSTACK-5297 is accepted with small enhancement in:
+            //KVMStorageProcessor::createVolumeFromSnapshot and CloudStackPrimaryDataStoreDriverImpl::copyAsync/createAsync
+            if ((!backupSnapToSecondary && (srcPool.getPoolType() == StoragePoolType.NetworkFilesystem || srcPool.getPoolType() == StoragePoolType.Filesystem))) {
+                SnapshotStrategy snapshotStrategy = _storageStrategyFactory.getSnapshotStrategy(snapshot, SnapshotOperation.BACKUP);
+                snapshotStrategy.backupSnapshot(snapInfo);
+                // Attempt to grab it again.
+                snapInfo = snapshotFactory.getSnapshot(snapshot.getId(), dataStoreRole);
+                if (snapInfo == null) {
+                    throw new CloudRuntimeException("Cannot find snapshot " + snapshot.getId() + " on secondary and could not create backup");
+                }
             }
         }
         // don't try to perform a sync if the DataStoreRole of the snapshot is equal to DataStoreRole.Primary
-        if (!DataStoreRole.Primary.equals(dataStoreRole)) {
+        if (!DataStoreRole.Primary.equals(snapInfo.getDataStore().getRole())) {
             try {
                 // sync snapshot to region store if necessary
                 DataStore snapStore = snapInfo.getDataStore();

--- a/engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java
+++ b/engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java
@@ -457,21 +457,7 @@ public class VolumeOrchestrator extends ManagerBase implements VolumeOrchestrati
                 throw new CloudRuntimeException("Cannot find snapshot " + snapshot.getId());
             }
 
-            boolean backupSnapToSecondary = SnapshotManager.BackupSnapshotAfterTakingSnapshot.value() == null || SnapshotManager.BackupSnapshotAfterTakingSnapshot.value();
-
-            StoragePoolVO srcPool = _storagePoolDao.findById(snapInfo.getBaseVolume().getPoolId());
-            // We need to copy the snapshot onto secondary.
-            //Skipping the backup to secondary storage with NFS/FS could be supported when CLOUDSTACK-5297 is accepted with small enhancement in:
-            //KVMStorageProcessor::createVolumeFromSnapshot and CloudStackPrimaryDataStoreDriverImpl::copyAsync/createAsync
-            if ((!backupSnapToSecondary && (srcPool.getPoolType() == StoragePoolType.NetworkFilesystem || srcPool.getPoolType() == StoragePoolType.Filesystem))) {
-                SnapshotStrategy snapshotStrategy = _storageStrategyFactory.getSnapshotStrategy(snapshot, SnapshotOperation.BACKUP);
-                snapshotStrategy.backupSnapshot(snapInfo);
-                // Attempt to grab it again.
-                snapInfo = snapshotFactory.getSnapshot(snapshot.getId(), dataStoreRole);
-                if (snapInfo == null) {
-                    throw new CloudRuntimeException("Cannot find snapshot " + snapshot.getId() + " on secondary and could not create backup");
-                }
-            }
+            snapInfo = backupSnapshotIfNeeded(snapshot, dataStoreRole, snapInfo);
         }
         // don't try to perform a sync if the DataStoreRole of the snapshot is equal to DataStoreRole.Primary
         if (!DataStoreRole.Primary.equals(snapInfo.getDataStore().getRole())) {
@@ -504,6 +490,25 @@ public class VolumeOrchestrator extends ManagerBase implements VolumeOrchestrati
             throw new CloudRuntimeException("Failed to create volume from snapshot", e);
         }
 
+    }
+
+    private SnapshotInfo backupSnapshotIfNeeded(Snapshot snapshot, DataStoreRole dataStoreRole, SnapshotInfo snapInfo) {
+        boolean backupSnapToSecondary = SnapshotManager.BackupSnapshotAfterTakingSnapshot.value() == null || SnapshotManager.BackupSnapshotAfterTakingSnapshot.value();
+
+        StoragePoolVO srcPool = _storagePoolDao.findById(snapInfo.getBaseVolume().getPoolId());
+        // We need to copy the snapshot onto secondary.
+        //Skipping the backup to secondary storage with NFS/FS could be supported when CLOUDSTACK-5297 is accepted with small enhancement in:
+        //KVMStorageProcessor::createVolumeFromSnapshot and CloudStackPrimaryDataStoreDriverImpl::copyAsync/createAsync
+        if ((!backupSnapToSecondary && (StoragePoolType.NetworkFilesystem.equals(srcPool.getPoolType()) || StoragePoolType.Filesystem.equals(srcPool.getPoolType())))) {
+            SnapshotStrategy snapshotStrategy = _storageStrategyFactory.getSnapshotStrategy(snapshot, SnapshotOperation.BACKUP);
+            snapshotStrategy.backupSnapshot(snapInfo);
+            // Attempt to grab it again.
+            snapInfo = snapshotFactory.getSnapshot(snapshot.getId(), dataStoreRole);
+            if (snapInfo == null) {
+                throw new CloudRuntimeException("Cannot find snapshot " + snapshot.getId() + " on secondary and could not create backup");
+            }
+        }
+        return snapInfo;
     }
 
     public DataStoreRole getDataStoreRole(Snapshot snapshot) {

--- a/engine/schema/src/main/java/org/apache/cloudstack/storage/datastore/db/SnapshotDataStoreDao.java
+++ b/engine/schema/src/main/java/org/apache/cloudstack/storage/datastore/db/SnapshotDataStoreDao.java
@@ -68,6 +68,8 @@ StateDao<ObjectInDataStoreStateMachine.State, ObjectInDataStoreStateMachine.Even
 
     SnapshotDataStoreVO findByVolume(long volumeId, DataStoreRole role);
 
+    SnapshotDataStoreVO findByVolume(long snapshotId, long volumeId, DataStoreRole role);
+
     /**
      * List all snapshots in 'snapshot_store_ref' by volume and data store role. Therefore, it is possible to list all snapshots that are in the primary storage or in the secondary storage.
      */

--- a/engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/SnapshotDataFactoryImpl.java
+++ b/engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/SnapshotDataFactoryImpl.java
@@ -92,7 +92,7 @@ public class SnapshotDataFactoryImpl implements SnapshotDataFactory {
         }
         SnapshotDataStoreVO snapshotStore = snapshotStoreDao.findBySnapshot(snapshotId, role);
         if (snapshotStore == null) {
-            snapshotStore = snapshotStoreDao.findByVolume(snapshot.getVolumeId(), role);
+            snapshotStore = snapshotStoreDao.findByVolume(snapshotId, snapshot.getVolumeId(), role);
             if (snapshotStore == null) {
                 return null;
             }

--- a/engine/storage/src/main/java/org/apache/cloudstack/storage/image/db/SnapshotDataStoreDaoImpl.java
+++ b/engine/storage/src/main/java/org/apache/cloudstack/storage/image/db/SnapshotDataStoreDaoImpl.java
@@ -142,6 +142,7 @@ public class SnapshotDataStoreDaoImpl extends GenericDaoBase<SnapshotDataStoreVO
         volumeSearch = createSearchBuilder();
         volumeSearch.and("volume_id", volumeSearch.entity().getVolumeId(), SearchCriteria.Op.EQ);
         volumeSearch.and("store_role", volumeSearch.entity().getRole(), SearchCriteria.Op.EQ);
+        volumeSearch.and("snapshot_id", volumeSearch.entity().getSnapshotId(), SearchCriteria.Op.EQ);
         volumeSearch.done();
 
         stateSearch = createSearchBuilder();
@@ -360,6 +361,15 @@ public class SnapshotDataStoreDaoImpl extends GenericDaoBase<SnapshotDataStoreVO
     @Override
     public SnapshotDataStoreVO findByVolume(long volumeId, DataStoreRole role) {
         SearchCriteria<SnapshotDataStoreVO> sc = volumeSearch.create();
+        sc.setParameters("volume_id", volumeId);
+        sc.setParameters("store_role", role);
+        return findOneBy(sc);
+    }
+
+    @Override
+    public SnapshotDataStoreVO findByVolume(long snapshotId, long volumeId, DataStoreRole role) {
+        SearchCriteria<SnapshotDataStoreVO> sc = volumeSearch.create();
+        sc.setParameters("snapshot_id", snapshotId);
         sc.setParameters("volume_id", volumeId);
         sc.setParameters("store_role", role);
         return findOneBy(sc);

--- a/plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java
+++ b/plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java
@@ -85,6 +85,7 @@ import com.ceph.rados.exceptions.RadosException;
 import com.ceph.rbd.Rbd;
 import com.ceph.rbd.RbdException;
 import com.ceph.rbd.RbdImage;
+import com.ceph.rbd.jna.RbdSnapInfo;
 import com.cloud.agent.api.Answer;
 import com.cloud.agent.api.storage.PrimaryStorageDownloadAnswer;
 import com.cloud.agent.api.to.DataObjectType;
@@ -1594,39 +1595,72 @@ public class KVMStorageProcessor implements StorageProcessor {
             final DataStoreTO imageStore = srcData.getDataStore();
             final VolumeObjectTO volume = snapshot.getVolume();
 
-            if (!(imageStore instanceof NfsTO)) {
+            if (!(imageStore instanceof NfsTO) && !(imageStore instanceof PrimaryDataStoreTO)) {
                 return new CopyCmdAnswer("unsupported protocol");
             }
-
-            final NfsTO nfsImageStore = (NfsTO)imageStore;
 
             final String snapshotFullPath = snapshot.getPath();
             final int index = snapshotFullPath.lastIndexOf("/");
             final String snapshotPath = snapshotFullPath.substring(0, index);
             final String snapshotName = snapshotFullPath.substring(index + 1);
-            final KVMStoragePool secondaryPool = storagePoolMgr.getStoragePoolByURI(nfsImageStore.getUrl() + File.separator + snapshotPath);
-            final KVMPhysicalDisk snapshotDisk = secondaryPool.getPhysicalDisk(snapshotName);
+            KVMPhysicalDisk snapshotDisk = null;
+            KVMPhysicalDisk disk = null;
+            if (imageStore instanceof NfsTO) {
+                final NfsTO nfsImageStore = (NfsTO)imageStore;
+                final KVMStoragePool secondaryPool = storagePoolMgr.getStoragePoolByURI(nfsImageStore.getUrl() + File.separator + snapshotPath);
+                snapshotDisk = secondaryPool.getPhysicalDisk(snapshotName);
+                if (volume.getFormat() == ImageFormat.RAW) {
+                    snapshotDisk.setFormat(PhysicalDiskFormat.RAW);
+                } else if (volume.getFormat() == ImageFormat.QCOW2) {
+                    snapshotDisk.setFormat(PhysicalDiskFormat.QCOW2);
+                }
 
-            if (volume.getFormat() == ImageFormat.RAW) {
-                snapshotDisk.setFormat(PhysicalDiskFormat.RAW);
-            } else if (volume.getFormat() == ImageFormat.QCOW2) {
-                snapshotDisk.setFormat(PhysicalDiskFormat.QCOW2);
+                final String primaryUuid = pool.getUuid();
+                final KVMStoragePool primaryPool = storagePoolMgr.getStoragePool(pool.getPoolType(), primaryUuid);
+                final String volUuid = UUID.randomUUID().toString();
+
+                Map<String, String> details = cmd.getOptions2();
+
+                String path = details != null ? details.get(DiskTO.IQN) : null;
+
+                storagePoolMgr.connectPhysicalDisk(pool.getPoolType(), pool.getUuid(), path, details);
+
+                disk = storagePoolMgr.copyPhysicalDisk(snapshotDisk, path != null ? path : volUuid, primaryPool, cmd.getWaitInMillSeconds());
+
+                storagePoolMgr.disconnectPhysicalDisk(pool.getPoolType(), pool.getUuid(), path);
+            } else {
+                final PrimaryDataStoreTO primaryStore = (PrimaryDataStoreTO) imageStore;
+                KVMStoragePool srcPool = storagePoolMgr.getStoragePool(primaryStore.getPoolType(), primaryStore.getUuid());
+                snapshotDisk = srcPool.getPhysicalDisk(volume.getPath());
+                KVMStoragePool destPool = storagePoolMgr.getStoragePool(pool.getPoolType(), pool.getUuid());
+                VolumeObjectTO newVol = (VolumeObjectTO) destData;
+                if (primaryStore.getPoolType() == StoragePoolType.RBD) {
+                    s_logger.debug(String.format("Attempting to create volume from RBD snapshot %s", snapshotName));
+                    if (pool.getPoolType() == StoragePoolType.RBD) {
+                        disk = createRBDvolumeFromRBDSnapshot(snapshotDisk, snapshotName, newVol.getUuid(),
+                                PhysicalDiskFormat.RAW, newVol.getSize(), destPool, cmd.getWaitInMillSeconds());
+                        s_logger.debug(String.format("Created RBD volume %s from snapshot %s", disk, snapshotDisk));
+                    } else {
+                        Map<String, String> details = cmd.getOptions2();
+
+                        String path = details != null ? details.get(DiskTO.IQN) : null;
+
+                        storagePoolMgr.connectPhysicalDisk(pool.getPoolType(), pool.getUuid(), path, details);
+
+                        snapshotDisk.setPath(snapshotDisk.getPath() + "@" + snapshotName);
+                        disk = storagePoolMgr.copyPhysicalDisk(snapshotDisk, path != null ? path : newVol.getUuid(),
+                                destPool, cmd.getWaitInMillSeconds());
+
+                        storagePoolMgr.disconnectPhysicalDisk(pool.getPoolType(), pool.getUuid(), path);
+                        s_logger.debug(String.format("Created RBD volume %s from snapshot %s", disk, snapshotDisk));
+
+                    }
+                }
             }
 
-            final String primaryUuid = pool.getUuid();
-            final KVMStoragePool primaryPool = storagePoolMgr.getStoragePool(pool.getPoolType(), primaryUuid);
-            final String volUuid = UUID.randomUUID().toString();
-
-            Map<String, String> details = cmd.getOptions2();
-
-            String path = details != null ? details.get(DiskTO.IQN) : null;
-
-            storagePoolMgr.connectPhysicalDisk(pool.getPoolType(), pool.getUuid(), path, details);
-
-            KVMPhysicalDisk disk = storagePoolMgr.copyPhysicalDisk(snapshotDisk, path != null ? path : volUuid, primaryPool, cmd.getWaitInMillSeconds());
-
-            storagePoolMgr.disconnectPhysicalDisk(pool.getPoolType(), pool.getUuid(), path);
-
+            if (disk == null) {
+                return new CopyCmdAnswer("Could not create volume from snapshot");
+            }
             final VolumeObjectTO newVol = new VolumeObjectTO();
             newVol.setPath(disk.getName());
             newVol.setSize(disk.getVirtualSize());
@@ -1637,6 +1671,74 @@ public class KVMStorageProcessor implements StorageProcessor {
             s_logger.debug("Failed to createVolumeFromSnapshot: ", e);
             return new CopyCmdAnswer(e.toString());
         }
+    }
+
+    private KVMPhysicalDisk createRBDvolumeFromRBDSnapshot(KVMPhysicalDisk volume, String snapshotName, String name,
+            PhysicalDiskFormat format, long size, KVMStoragePool destPool, int timeout) {
+
+        KVMStoragePool srcPool = volume.getPool();
+        KVMPhysicalDisk disk = null;
+        String newUuid = name;
+        int rbdFeatures = 61;
+
+        format = PhysicalDiskFormat.RAW;
+        disk = new KVMPhysicalDisk(destPool.getSourceDir() + "/" + newUuid, newUuid, destPool);
+        disk.setFormat(format);
+        if (size > volume.getVirtualSize()) {
+            disk.setSize(size);
+            disk.setVirtualSize(size);
+        } else {
+            disk.setSize(volume.getVirtualSize());
+            disk.setVirtualSize(disk.getSize());
+        }
+
+        try {
+
+            Rados r = new Rados(srcPool.getAuthUserName());
+            r.confSet("mon_host", srcPool.getSourceHost() + ":" + srcPool.getSourcePort());
+            r.confSet("key", srcPool.getAuthSecret());
+            r.confSet("client_mount_timeout", "30");
+            r.connect();
+
+            IoCTX io = r.ioCtxCreate(srcPool.getSourceDir());
+            Rbd rbd = new Rbd(io);
+            RbdImage srcImage = rbd.open(volume.getName());
+
+            List<RbdSnapInfo> snaps = srcImage.snapList();
+            boolean snapFound = false;
+            for (RbdSnapInfo snap : snaps) {
+                if (snapshotName.equals(snap.name)) {
+                    snapFound = true;
+                    break;
+                }
+            }
+
+            if (!snapFound) {
+                return null;
+            }
+            srcImage.snapProtect(snapshotName);
+
+            rbd.clone(volume.getName(), snapshotName, io, disk.getName(), rbdFeatures, 0);
+            RbdImage diskImage = rbd.open(disk.getName());
+            if (disk.getVirtualSize() > volume.getVirtualSize()) {
+                diskImage.resize(disk.getVirtualSize());
+            }
+
+            diskImage.flatten();
+            rbd.close(diskImage);
+
+            srcImage.snapUnprotect(snapshotName);
+            rbd.close(srcImage);
+            r.ioCtxDestroy(io);
+        } catch (RadosException e) {
+            s_logger.error(String.format("Failed due to %s", e.getMessage()));
+            disk = null;
+        } catch (RbdException e) {
+            s_logger.error(String.format("Failed due to: %s", e.getMessage()));
+            disk = null;
+        }
+
+        return disk;
     }
 
     @Override

--- a/plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java
+++ b/plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java
@@ -1595,7 +1595,7 @@ public class KVMStorageProcessor implements StorageProcessor {
             final DataStoreTO imageStore = srcData.getDataStore();
             final VolumeObjectTO volume = snapshot.getVolume();
 
-            if (!(imageStore instanceof NfsTO) && !(imageStore instanceof PrimaryDataStoreTO)) {
+            if (!(imageStore instanceof NfsTO || imageStore instanceof PrimaryDataStoreTO)) {
                 return new CopyCmdAnswer("unsupported protocol");
             }
 
@@ -1603,59 +1603,11 @@ public class KVMStorageProcessor implements StorageProcessor {
             final int index = snapshotFullPath.lastIndexOf("/");
             final String snapshotPath = snapshotFullPath.substring(0, index);
             final String snapshotName = snapshotFullPath.substring(index + 1);
-            KVMPhysicalDisk snapshotDisk = null;
             KVMPhysicalDisk disk = null;
             if (imageStore instanceof NfsTO) {
-                final NfsTO nfsImageStore = (NfsTO)imageStore;
-                final KVMStoragePool secondaryPool = storagePoolMgr.getStoragePoolByURI(nfsImageStore.getUrl() + File.separator + snapshotPath);
-                snapshotDisk = secondaryPool.getPhysicalDisk(snapshotName);
-                if (volume.getFormat() == ImageFormat.RAW) {
-                    snapshotDisk.setFormat(PhysicalDiskFormat.RAW);
-                } else if (volume.getFormat() == ImageFormat.QCOW2) {
-                    snapshotDisk.setFormat(PhysicalDiskFormat.QCOW2);
-                }
-
-                final String primaryUuid = pool.getUuid();
-                final KVMStoragePool primaryPool = storagePoolMgr.getStoragePool(pool.getPoolType(), primaryUuid);
-                final String volUuid = UUID.randomUUID().toString();
-
-                Map<String, String> details = cmd.getOptions2();
-
-                String path = details != null ? details.get(DiskTO.IQN) : null;
-
-                storagePoolMgr.connectPhysicalDisk(pool.getPoolType(), pool.getUuid(), path, details);
-
-                disk = storagePoolMgr.copyPhysicalDisk(snapshotDisk, path != null ? path : volUuid, primaryPool, cmd.getWaitInMillSeconds());
-
-                storagePoolMgr.disconnectPhysicalDisk(pool.getPoolType(), pool.getUuid(), path);
+                disk = createVolumeFromSnapshotOnNFS(cmd, pool, imageStore, volume, snapshotPath, snapshotName);
             } else {
-                final PrimaryDataStoreTO primaryStore = (PrimaryDataStoreTO) imageStore;
-                KVMStoragePool srcPool = storagePoolMgr.getStoragePool(primaryStore.getPoolType(), primaryStore.getUuid());
-                snapshotDisk = srcPool.getPhysicalDisk(volume.getPath());
-                KVMStoragePool destPool = storagePoolMgr.getStoragePool(pool.getPoolType(), pool.getUuid());
-                VolumeObjectTO newVol = (VolumeObjectTO) destData;
-                if (primaryStore.getPoolType() == StoragePoolType.RBD) {
-                    s_logger.debug(String.format("Attempting to create volume from RBD snapshot %s", snapshotName));
-                    if (pool.getPoolType() == StoragePoolType.RBD) {
-                        disk = createRBDvolumeFromRBDSnapshot(snapshotDisk, snapshotName, newVol.getUuid(),
-                                PhysicalDiskFormat.RAW, newVol.getSize(), destPool, cmd.getWaitInMillSeconds());
-                        s_logger.debug(String.format("Created RBD volume %s from snapshot %s", disk, snapshotDisk));
-                    } else {
-                        Map<String, String> details = cmd.getOptions2();
-
-                        String path = details != null ? details.get(DiskTO.IQN) : null;
-
-                        storagePoolMgr.connectPhysicalDisk(pool.getPoolType(), pool.getUuid(), path, details);
-
-                        snapshotDisk.setPath(snapshotDisk.getPath() + "@" + snapshotName);
-                        disk = storagePoolMgr.copyPhysicalDisk(snapshotDisk, path != null ? path : newVol.getUuid(),
-                                destPool, cmd.getWaitInMillSeconds());
-
-                        storagePoolMgr.disconnectPhysicalDisk(pool.getPoolType(), pool.getUuid(), path);
-                        s_logger.debug(String.format("Created RBD volume %s from snapshot %s", disk, snapshotDisk));
-
-                    }
-                }
+                disk = createVolumeFromRBDSnapshot(cmd, destData, pool, imageStore, volume, snapshotName, disk);
             }
 
             if (disk == null) {
@@ -1673,24 +1625,77 @@ public class KVMStorageProcessor implements StorageProcessor {
         }
     }
 
+    private KVMPhysicalDisk createVolumeFromRBDSnapshot(CopyCommand cmd, DataTO destData,
+            PrimaryDataStoreTO pool, DataStoreTO imageStore, VolumeObjectTO volume, String snapshotName, KVMPhysicalDisk disk) {
+        PrimaryDataStoreTO primaryStore = (PrimaryDataStoreTO) imageStore;
+        KVMStoragePool srcPool = storagePoolMgr.getStoragePool(primaryStore.getPoolType(), primaryStore.getUuid());
+        KVMPhysicalDisk snapshotDisk = srcPool.getPhysicalDisk(volume.getPath());
+        KVMStoragePool destPool = storagePoolMgr.getStoragePool(pool.getPoolType(), pool.getUuid());
+        VolumeObjectTO newVol = (VolumeObjectTO) destData;
+
+        if (StoragePoolType.RBD.equals(primaryStore.getPoolType())) {
+            s_logger.debug(String.format("Attempting to create volume from RBD snapshot %s", snapshotName));
+            if (StoragePoolType.RBD.equals(pool.getPoolType())) {
+                disk = createRBDvolumeFromRBDSnapshot(snapshotDisk, snapshotName, newVol.getUuid(),
+                        PhysicalDiskFormat.RAW, newVol.getSize(), destPool, cmd.getWaitInMillSeconds());
+                s_logger.debug(String.format("Created RBD volume %s from snapshot %s", disk, snapshotDisk));
+            } else {
+                Map<String, String> details = cmd.getOptions2();
+
+                String path = details != null ? details.get(DiskTO.IQN) : null;
+
+                storagePoolMgr.connectPhysicalDisk(pool.getPoolType(), pool.getUuid(), path, details);
+
+                snapshotDisk.setPath(snapshotDisk.getPath() + "@" + snapshotName);
+                disk = storagePoolMgr.copyPhysicalDisk(snapshotDisk, path != null ? path : newVol.getUuid(),
+                        destPool, cmd.getWaitInMillSeconds());
+
+                storagePoolMgr.disconnectPhysicalDisk(pool.getPoolType(), pool.getUuid(), path);
+                s_logger.debug(String.format("Created RBD volume %s from snapshot %s", disk, snapshotDisk));
+
+            }
+        }
+        return disk;
+    }
+
+    private KVMPhysicalDisk createVolumeFromSnapshotOnNFS(CopyCommand cmd, PrimaryDataStoreTO pool,
+            DataStoreTO imageStore, VolumeObjectTO volume, String snapshotPath, String snapshotName) {
+        NfsTO nfsImageStore = (NfsTO)imageStore;
+        KVMStoragePool secondaryPool = storagePoolMgr.getStoragePoolByURI(nfsImageStore.getUrl() + File.separator + snapshotPath);
+        KVMPhysicalDisk snapshotDisk = secondaryPool.getPhysicalDisk(snapshotName);
+        if (volume.getFormat() == ImageFormat.RAW) {
+            snapshotDisk.setFormat(PhysicalDiskFormat.RAW);
+        } else if (volume.getFormat() == ImageFormat.QCOW2) {
+            snapshotDisk.setFormat(PhysicalDiskFormat.QCOW2);
+        }
+
+        final String primaryUuid = pool.getUuid();
+        final KVMStoragePool primaryPool = storagePoolMgr.getStoragePool(pool.getPoolType(), primaryUuid);
+        final String volUuid = UUID.randomUUID().toString();
+
+        Map<String, String> details = cmd.getOptions2();
+
+        String path = details != null ? details.get(DiskTO.IQN) : null;
+
+        storagePoolMgr.connectPhysicalDisk(pool.getPoolType(), pool.getUuid(), path, details);
+
+        KVMPhysicalDisk disk = storagePoolMgr.copyPhysicalDisk(snapshotDisk, path != null ? path : volUuid, primaryPool, cmd.getWaitInMillSeconds());
+
+        storagePoolMgr.disconnectPhysicalDisk(pool.getPoolType(), pool.getUuid(), path);
+        return disk;
+    }
+
     private KVMPhysicalDisk createRBDvolumeFromRBDSnapshot(KVMPhysicalDisk volume, String snapshotName, String name,
             PhysicalDiskFormat format, long size, KVMStoragePool destPool, int timeout) {
 
         KVMStoragePool srcPool = volume.getPool();
         KVMPhysicalDisk disk = null;
         String newUuid = name;
-        int rbdFeatures = 61;
 
-        format = PhysicalDiskFormat.RAW;
         disk = new KVMPhysicalDisk(destPool.getSourceDir() + "/" + newUuid, newUuid, destPool);
         disk.setFormat(format);
-        if (size > volume.getVirtualSize()) {
-            disk.setSize(size);
-            disk.setVirtualSize(size);
-        } else {
-            disk.setSize(volume.getVirtualSize());
-            disk.setVirtualSize(disk.getSize());
-        }
+        disk.setSize(size > volume.getVirtualSize() ? size : volume.getVirtualSize());
+        disk.setVirtualSize(size > volume.getVirtualSize() ? size : disk.getSize());
 
         try {
 
@@ -1714,11 +1719,13 @@ public class KVMStorageProcessor implements StorageProcessor {
             }
 
             if (!snapFound) {
+                s_logger.debug(String.format("Could not find snapshot %s on RBD", snapshotName));
                 return null;
             }
             srcImage.snapProtect(snapshotName);
 
-            rbd.clone(volume.getName(), snapshotName, io, disk.getName(), rbdFeatures, 0);
+            s_logger.debug(String.format("Try to clone snapshot %s on RBD", snapshotName));
+            rbd.clone(volume.getName(), snapshotName, io, disk.getName(), LibvirtStorageAdaptor.RBD_FEATURES, 0);
             RbdImage diskImage = rbd.open(disk.getName());
             if (disk.getVirtualSize() > volume.getVirtualSize()) {
                 diskImage.resize(disk.getVirtualSize());
@@ -1730,11 +1737,8 @@ public class KVMStorageProcessor implements StorageProcessor {
             srcImage.snapUnprotect(snapshotName);
             rbd.close(srcImage);
             r.ioCtxDestroy(io);
-        } catch (RadosException e) {
-            s_logger.error(String.format("Failed due to %s", e.getMessage()));
-            disk = null;
-        } catch (RbdException e) {
-            s_logger.error(String.format("Failed due to: %s", e.getMessage()));
+        } catch (RadosException | RbdException e) {
+            s_logger.error(String.format("Failed due to %s", e.getMessage()), e);
             disk = null;
         }
 

--- a/plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/LibvirtStorageAdaptor.java
+++ b/plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/LibvirtStorageAdaptor.java
@@ -79,7 +79,7 @@ public class LibvirtStorageAdaptor implements StorageAdaptor {
     private static final int RBD_FEATURE_OBJECT_MAP = 8;
     private static final int RBD_FEATURE_FAST_DIFF = 16;
     private static final int RBD_FEATURE_DEEP_FLATTEN = 32;
-    private int rbdFeatures = RBD_FEATURE_LAYERING + RBD_FEATURE_EXCLUSIVE_LOCK + RBD_FEATURE_OBJECT_MAP + RBD_FEATURE_FAST_DIFF + RBD_FEATURE_DEEP_FLATTEN;
+    public static final int RBD_FEATURES = RBD_FEATURE_LAYERING + RBD_FEATURE_EXCLUSIVE_LOCK + RBD_FEATURE_OBJECT_MAP + RBD_FEATURE_FAST_DIFF + RBD_FEATURE_DEEP_FLATTEN;
     private int rbdOrder = 0; /* Order 0 means 4MB blocks (the default) */
 
     public LibvirtStorageAdaptor(StorageLayer storage) {
@@ -1103,7 +1103,7 @@ public class LibvirtStorageAdaptor implements StorageAdaptor {
                         s_logger.debug("The source image " + srcPool.getSourceDir() + "/" + template.getName() +
                                 " is RBD format 1. We have to perform a regular copy (" + toHumanReadableSize(disk.getVirtualSize()) + " bytes)");
 
-                        rbd.create(disk.getName(), disk.getVirtualSize(), rbdFeatures, rbdOrder);
+                        rbd.create(disk.getName(), disk.getVirtualSize(), RBD_FEATURES, rbdOrder);
                         RbdImage destImage = rbd.open(disk.getName());
 
                         s_logger.debug("Starting to copy " + srcImage.getName() +  " to " + destImage.getName() + " in Ceph pool " + srcPool.getSourceDir());
@@ -1140,7 +1140,7 @@ public class LibvirtStorageAdaptor implements StorageAdaptor {
                             srcImage.snapProtect(rbdTemplateSnapName);
                         }
 
-                        rbd.clone(template.getName(), rbdTemplateSnapName, io, disk.getName(), rbdFeatures, rbdOrder);
+                        rbd.clone(template.getName(), rbdTemplateSnapName, io, disk.getName(), RBD_FEATURES, rbdOrder);
                         s_logger.debug("Succesfully cloned " + template.getName() + "@" + rbdTemplateSnapName + " to " + disk.getName());
                         /* We also need to resize the image if the VM was deployed with a larger root disk size */
                         if (disk.getVirtualSize() > template.getVirtualSize()) {
@@ -1180,7 +1180,7 @@ public class LibvirtStorageAdaptor implements StorageAdaptor {
 
                     s_logger.debug("Creating " + disk.getName() + " on the destination cluster " + rDest.confGet("mon_host") + " in pool " +
                             destPool.getSourceDir());
-                    dRbd.create(disk.getName(), disk.getVirtualSize(), rbdFeatures, rbdOrder);
+                    dRbd.create(disk.getName(), disk.getVirtualSize(), RBD_FEATURES, rbdOrder);
 
                     RbdImage srcImage = sRbd.open(template.getName());
                     RbdImage destImage = dRbd.open(disk.getName());

--- a/plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/driver/CloudStackPrimaryDataStoreDriverImpl.java
+++ b/plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/driver/CloudStackPrimaryDataStoreDriverImpl.java
@@ -316,14 +316,14 @@ public class CloudStackPrimaryDataStoreDriverImpl implements PrimaryDataStoreDri
             if (storagePoolVO != null && storagePoolVO.getPoolType() == Storage.StoragePoolType.CLVM) {
                 return true;
             }
-        } else if (srcData.getType() == DataObjectType.SNAPSHOT && destData.getType() == DataObjectType.VOLUME) {
+        } else if (DataObjectType.SNAPSHOT.equals(srcData.getType()) && DataObjectType.VOLUME.equals(destData.getType())) {
             DataStore srcStore = srcData.getDataStore();
-            if (srcStore.getRole() == DataStoreRole.Primary && destStore.getRole() == DataStoreRole.Primary) {
+            if (DataStoreRole.Primary.equals(srcStore.getRole()) && DataStoreRole.Primary.equals(destStore.getRole())) {
                 StoragePoolVO srcStoragePoolVO = primaryStoreDao.findById(srcStore.getId());
                 StoragePoolVO dstStoragePoolVO = primaryStoreDao.findById(destStore.getId());
-                if (srcStoragePoolVO != null && srcStoragePoolVO.getPoolType() == StoragePoolType.RBD
-                        && dstStoragePoolVO != null && (dstStoragePoolVO.getPoolType() == StoragePoolType.RBD
-                        || dstStoragePoolVO.getPoolType() == StoragePoolType.NetworkFilesystem)) {
+                if (srcStoragePoolVO != null && StoragePoolType.RBD.equals(srcStoragePoolVO.getPoolType())
+                        && dstStoragePoolVO != null && (StoragePoolType.RBD.equals(dstStoragePoolVO.getPoolType())
+                        || StoragePoolType.NetworkFilesystem.equals(dstStoragePoolVO.getPoolType()))) {
                     return true;
                 }
             }

--- a/plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/driver/CloudStackPrimaryDataStoreDriverImpl.java
+++ b/plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/driver/CloudStackPrimaryDataStoreDriverImpl.java
@@ -69,6 +69,7 @@ import com.cloud.storage.CreateSnapshotPayload;
 import com.cloud.storage.DataStoreRole;
 import com.cloud.storage.ResizeVolumePayload;
 import com.cloud.storage.Storage;
+import com.cloud.storage.Storage.StoragePoolType;
 import com.cloud.storage.StorageManager;
 import com.cloud.storage.StoragePool;
 import com.cloud.storage.Volume;
@@ -287,6 +288,20 @@ public class CloudStackPrimaryDataStoreDriverImpl implements PrimaryDataStoreDri
                 }
                 CopyCommandResult result = new CopyCommandResult("", answer);
                 callback.complete(result);
+            } else if (srcdata.getType() == DataObjectType.SNAPSHOT && destData.getType() == DataObjectType.VOLUME) {
+                SnapshotObjectTO srcTO = (SnapshotObjectTO) srcdata.getTO();
+                CopyCommand cmd = new CopyCommand(srcTO, destData.getTO(), StorageManager.PRIMARY_STORAGE_DOWNLOAD_WAIT.value(), true);
+                EndPoint ep = epSelector.select(srcdata, destData);
+                CopyCmdAnswer answer = null;
+                if (ep == null) {
+                    String errMsg = "No remote endpoint to send command, check if host or ssvm is down?";
+                    s_logger.error(errMsg);
+                    answer = new CopyCmdAnswer(errMsg);
+                } else {
+                    answer = (CopyCmdAnswer) ep.sendMessage(cmd);
+                }
+                CopyCommandResult result = new CopyCommandResult("", answer);
+                callback.complete(result);
             }
         }
     }
@@ -294,12 +309,23 @@ public class CloudStackPrimaryDataStoreDriverImpl implements PrimaryDataStoreDri
     @Override
     public boolean canCopy(DataObject srcData, DataObject destData) {
         //BUG fix for CLOUDSTACK-4618
-        DataStore store = destData.getDataStore();
-        if (store.getRole() == DataStoreRole.Primary && srcData.getType() == DataObjectType.TEMPLATE
+        DataStore destStore = destData.getDataStore();
+        if (destStore.getRole() == DataStoreRole.Primary && srcData.getType() == DataObjectType.TEMPLATE
                 && (destData.getType() == DataObjectType.TEMPLATE || destData.getType() == DataObjectType.VOLUME)) {
-            StoragePoolVO storagePoolVO = primaryStoreDao.findById(store.getId());
+            StoragePoolVO storagePoolVO = primaryStoreDao.findById(destStore.getId());
             if (storagePoolVO != null && storagePoolVO.getPoolType() == Storage.StoragePoolType.CLVM) {
                 return true;
+            }
+        } else if (srcData.getType() == DataObjectType.SNAPSHOT && destData.getType() == DataObjectType.VOLUME) {
+            DataStore srcStore = srcData.getDataStore();
+            if (srcStore.getRole() == DataStoreRole.Primary && destStore.getRole() == DataStoreRole.Primary) {
+                StoragePoolVO srcStoragePoolVO = primaryStoreDao.findById(srcStore.getId());
+                StoragePoolVO dstStoragePoolVO = primaryStoreDao.findById(destStore.getId());
+                if (srcStoragePoolVO != null && srcStoragePoolVO.getPoolType() == StoragePoolType.RBD
+                        && dstStoragePoolVO != null && (dstStoragePoolVO.getPoolType() == StoragePoolType.RBD
+                        || dstStoragePoolVO.getPoolType() == StoragePoolType.NetworkFilesystem)) {
+                    return true;
+                }
             }
         }
         return false;

